drivers: wifi: siwx91x: Add support of disconnect result event

Nitin-Pandey-01 opened the pull request:
- Added support for disconnect result event
- Handling error cases in Wi-Fi disconnect

jhedberg commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -121,6 +134,58 @@ static enum wifi_mfp_options siwx91x_set_sta_mfp_option(sl_wifi_security_t secur
+
+	switch (module_state) {
+	case STATE_BEACON_LOSS:
+		LOG_ERR("Beacon Loss");
+		LOG_ERR("Reason: %s", reason_str);
```
I'm not sure what the policy of logging levels is in wifi drivers, but at least in Bluetooth we use `LOG_WRN()` for situations where some external state is unexpected or we receive unexpected/invalid data from an external entity. `LOG_ERR()` OTOH is reserved for internal/local errors.

Nitin-Pandey-01 replied:
Updated the code.

jhedberg commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -121,6 +134,58 @@ static enum wifi_mfp_options siwx91x_set_sta_mfp_option(sl_wifi_security_t secur
+		LOG_INF("Reason: %s", reason_str);
+		break;
+	case STATE_BETTER_AP_FOUND:
+		LOG_INF("Better AP found while roaming");
+		LOG_INF("Reason: %s", reason_str);
```
`LOG_DBG()` perhaps?

Nitin-Pandey-01 replied:
Sure

jhedberg commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -121,6 +134,58 @@ static enum wifi_mfp_options siwx91x_set_sta_mfp_option(sl_wifi_security_t secur
+	case STATE_DEAUTHENTICATION:
+		sidev->state = WIFI_STATE_DISCONNECTED;
+		wifi_mgmt_raise_disconnect_result_event(sidev->iface, WIFI_REASON_DISCONN_SUCCESS);
+		LOG_INF("Reason: %s", reason_str);
```
`LOG_DBG()`?

Nitin-Pandey-01 replied:
Sure

jhedberg commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -121,6 +134,58 @@ static enum wifi_mfp_options siwx91x_set_sta_mfp_option(sl_wifi_security_t secur
+	case STATE_ASSOCIATED:
+		sidev->state = WIFI_STATE_COMPLETED;
+		if (FIELD_GET(SIWX91X_INTERFACE_MASK, interface) == SL_WIFI_CLIENT_INTERFACE) {
+			LOG_INF("Associated");
```
Same here. In general, I think it's better to keep `LOG_INF()` use to the bare minimum to avoid spamming the console.

Nitin-Pandey-01 replied:
Ok, Updated the code as suggested

jhedberg commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -121,6 +134,58 @@ static enum wifi_mfp_options siwx91x_set_sta_mfp_option(sl_wifi_security_t secur
+			sidev->state = WIFI_STATE_DISCONNECTED;
+			wifi_mgmt_raise_disconnect_result_event(sidev->iface,
+								WIFI_REASON_DISCONN_SUCCESS);
+			LOG_INF("Reason: %s", reason_str);
```
Here as well.

Nitin-Pandey-01 replied:
Done

jhedberg reviewed (commented):
Some log level related comments

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -121,6 +134,58 @@ static enum wifi_mfp_options siwx91x_set_sta_mfp_option(sl_wifi_security_t secur
+{
+	struct siwx91x_dev *sidev = arg;
+	sl_si91x_module_state_stats_response_t *notif =
+		(sl_si91x_module_state_stats_response_t *)response;
```
I believe you don't need to cast the value.

Nitin-Pandey-01 replied:
Ok

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -369,6 +434,8 @@ static int siwx91x_connect(const struct device *dev, struct wifi_connect_req_par
 	if (params->channel != WIFI_CHANNEL_ANY) {
 		wifi_config.channel.channel = params->channel;
+	} else {
+		wifi_config.channel.channel = 0;
```
`wifi_config.channel.channel` is already 0.

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -385,16 +452,31 @@ static int siwx91x_connect(const struct device *dev, struct wifi_connect_req_par
-	int ret;
+	int ret = 0;
+
+	sl_wifi_interface_t interface = sl_wifi_get_default_interface();
```
Keep this declaration with the other ones (and sort them in the revert christmas tree order).

Nitin-Pandey-01 replied:
Done

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -385,16 +452,31 @@ static int siwx91x_connect(const struct device *dev, struct wifi_connect_req_par
 {
 	struct siwx91x_dev *sidev = dev->data;
-	int ret;
+	int ret = 0;
```
`ret` is going to be overwritten anyway, so don't initialize it.

Nitin-Pandey-01 replied:
Ok Sure

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -385,16 +452,31 @@ static int siwx91x_connect(const struct device *dev, struct wifi_connect_req_par
+	if (sidev->state != WIFI_STATE_COMPLETED) {
+		LOG_ERR("Command given in invalid state");
+		return -EBUSY;
+	}
```
We have to wait to join the network before to be able to disconnect it? What's pity.

Nitin-Pandey-01 replied:
This is just for the case to ensure that device is in proper state before sending disconnect command to firmware because SDK doesn't check this, What do you suggest ?

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi.h:
```diff
@@ -14,6 +14,25 @@
 #include "sl_si91x_types.h"
 #include "sl_si91x_protocol_types.h"
 
+typedef uint8_t sl_si91x_wlan_module_state_t;
```
Orphan?

Nitin-Pandey-01 replied:
Removed

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -33,12 +33,14 @@ int siwg91x_get_nwp_config(int wifi_oper_mode, sl_wifi_device_configuration_t *g
 			.ext_custom_feature_bit_map = SL_SI91X_EXT_FEAT_XTAL_CLK,
-		}
-	};
+		}};
```
Keep original alignement.

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -33,12 +33,14 @@ int siwg91x_get_nwp_config(int wifi_oper_mode, sl_wifi_device_configuration_t *g
 			.tcp_ip_feature_bit_map = SL_SI91X_TCP_IP_FEAT_EXTENSION_VALID,
-			.custom_feature_bit_map = SL_SI91X_CUSTOM_FEAT_EXTENSION_VALID,
+			.custom_feature_bit_map = (SL_SI91X_CUSTOM_FEAT_EXTENSION_VALID |
+						   SL_SI91X_CUSTOM_FEAT_ASYNC_CONNECTION_STATUS),
```
Remove useless parentheses.

Nitin-Pandey-01 replied:
Done

jerome-pouiller commented on soc/silabs/silabs_siwx91x/siwg917/nwp.c:
```diff
@@ -33,12 +33,14 @@ int siwg91x_get_nwp_config(int wifi_oper_mode, sl_wifi_device_configuration_t *g
-			.feature_bit_map = SL_SI91X_FEAT_SECURITY_OPEN | SL_SI91X_FEAT_WPS_DISABLE,
+			.feature_bit_map = SL_SI91X_FEAT_SECURITY_PSK | SL_SI91X_FEAT_WPS_DISABLE |
+					   SL_SI91X_FEAT_AGGREGATION |
+					   SL_SI91X_FEAT_HIDE_PSK_CREDENTIALS,
```
This change is unrelated. Place it in another commit.

BTW, does it make sense to enable these option even if Wifi is not enabled?

Nitin-Pandey-01 replied:
Ok, shifted this to another commit.

These features are specific to WiFi functionality, so enabling them when WiFi is not enabled might not make sense. We could consider adding conditional checks like IS_ENABLED(CONFIG_WIFI_SILABS_SIWX91X) to ensure they are only enabled when WiFi is active, What do you suggest ?

jerome-pouiller reviewed (commented):
Can you also address the comments Johan made?

Nitin-Pandey-01 commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -28,6 +28,19 @@ LOG_MODULE_REGISTER(siwx91x_wifi);
 
 NET_BUF_POOL_FIXED_DEFINE(siwx91x_tx_pool, 1, _NET_ETH_MAX_FRAME_SIZE, 0, NULL);
 
+static const struct siwx91x_bit_to_string REASONCODES[WLAN_REASON_CODES] = {
```
Removed the useless code and converted the structure to function.

Nitin-Pandey-01 commented on drivers/wifi/siwx91x/siwx91x_wifi.h:
```diff
@@ -14,6 +14,25 @@
+struct siwx91x_bit_to_string {
+	uint8_t bit;
+	char *string;
+};
```
Removed the definition

nipandey-01 committed:
driver: wifi: siwx91x: Update boot config

- Enabled WiFi-supported features in the feature bitmap.

Signed-off-by: Nitin Pandey <nitin.pandey@silabs.com>

nipandey-01 committed:
driver: wifi: siwx91x: Add support of disconnect result event

- Added support for disconnect result event
- Handling error cases in Wi-Fi disconnect

Signed-off-by: Nitin Pandey <nitin.pandey@silabs.com>

jhedberg commented on drivers/wifi/siwx91x/siwx91x_wifi.c:
```diff
@@ -28,6 +28,34 @@ LOG_MODULE_REGISTER(siwx91x_wifi);
+	static const struct {
+		int code;
+		const char *str;
+	} reason_strings[10] = {
```
minor: I'd remove the `10` and keep the size implicit based on the number of entries you define. Otherwise you need to remember to update this number if you ever add or remove something, which in turn is a potential source of bugs.

Nitin-Pandey-01 replied:
Indeed! But I got a Sonarqube error to define the size of the array.

jhedberg replied:
AFAIK SonarQube reports are not binding (they don't fail CI, do they?), and I'd disagree with it in this case (for the reasons I outlined in my comment). It might also be possible to tweak its configuration not to report this. @nashif any comments in this?

jhedberg replied:
@Nitin-Pandey-01 FYI: https://github.com/zephyrproject-rtos/zephyr/issues/89724#issuecomment-2866090088

jerome-pouiller commented:
I believe this PR is superseded by https://github.com/zephyrproject-rtos/zephyr/pull/87570. If it is the case, can you close this one @Nitin-Pandey-01?

Nitin-Pandey-01 commented:
Yes, that is true, Hence, I am closing this PR
